// typia@7.2.0/lib/functional.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertFunction = assertFunction;
exports.assertParameters = assertParameters;
exports.assertReturn = assertReturn;
exports.assertEqualsFunction = assertEqualsFunction;
exports.assertEqualsParameters = assertEqualsParameters;
exports.assertEqualsReturn = assertEqualsReturn;
exports.isFunction = isFunction;
exports.isParameters = isParameters;
exports.isReturn = isReturn;
exports.equalsFunction = equalsFunction;
exports.equalsParameters = equalsParameters;
exports.equalsReturn = equalsReturn;
exports.validateFunction = validateFunction;
exports.validateParameters = validateParameters;
exports.validateReturn = validateReturn;
exports.validateEqualsFunction = validateEqualsFunction;
exports.validateEqualsParameters = validateEqualsParameters;
exports.validateEqualsReturn = validateEqualsReturn;
/**
 * @internal
 */
function assertFunction() {
    halt("assertFunction");
}
/**
 * @internal
 */
function assertParameters() {
    halt("assertParameters");
}
/**
 * @internal
 */
function assertReturn() {
    halt("assertReturn");
}
/**
 * @internal
 */
function assertEqualsFunction() {
    halt("assertEqualsFunction");
}
/**
 * @internal
 */
function assertEqualsParameters() {
    halt("assertEqualsParameters");
}
/**
 * @internal
 */
function assertEqualsReturn() {
    halt("assertEqualsReturn");
}
/**
 * @internal
 */
function isFunction() {
    halt("isFunction");
}
/**
 * @internal
 */
function isParameters() {
    halt("isParameters");
}
/**
 * @internal
 */
function isReturn() {
    halt("isReturn");
}
/**
 * @internal
 */
function equalsFunction() {
    halt("equalsFunction");
}
/**
 * @internal
 */
function equalsParameters() {
    halt("equalsParameters");
}
/**
 * @internal
 */
function equalsReturn() {
    halt("equalsReturn");
}
/**
 * @internal
 */
function validateFunction() {
    halt("validateFunction");
}
/**
 * @internal
 */
function validateParameters() {
    halt("validateReturn");
}
/**
 * @internal
 */
function validateReturn() {
    halt("validateReturn");
}
/**
 * @internal
 */
function validateEqualsFunction() {
    halt("validateEqualsFunction");
}
/**
 * @internal
 */
function validateEqualsParameters() {
    halt("validateEqualsParameters");
}
/**
 * @internal
 */
function validateEqualsReturn() {
    halt("validateEqualsReturn");
}
/* -----------------------------------------------------------
  HALTER
----------------------------------------------------------- */
/**
 * @internal
 */
function halt(name) {
    throw new Error(`Error on typia.functional.${name}(): no transform has been configured. Read and follow https://typia.io/docs/setup please.`);
}
