// typia@7.2.0/lib/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemas = schemas;
exports.application = application;
exports.assertParse = assertParse;
exports.isParse = isParse;
exports.validateParse = validateParse;
exports.stringify = stringify;
exports.assertStringify = assertStringify;
exports.isStringify = isStringify;
exports.validateStringify = validateStringify;
exports.createIsParse = createIsParse;
exports.createAssertParse = createAssertParse;
exports.createValidateParse = createValidateParse;
exports.createStringify = createStringify;
exports.createAssertStringify = createAssertStringify;
exports.createIsStringify = createIsStringify;
exports.createValidateStringify = createValidateStringify;
/**
 * @internal
 */
function schemas() {
    halt("schemas");
}
/**
 * @internal
 */
function application() {
    halt("application");
}
/**
 * @internal
 */
function assertParse() {
    halt("assertParse");
}
/**
 * @internal
 */
function isParse() {
    halt("isParse");
}
/**
 * @internal
 */
function validateParse() {
    halt("validateParse");
}
/**
 * @internal
 */
function stringify() {
    halt("stringify");
}
/**
 * @internal
 */
function assertStringify() {
    halt("assertStringify");
}
/**
 * @internal
 */
function isStringify() {
    halt("isStringify");
}
/**
 * @internal
 */
function validateStringify() {
    halt("validateStringify");
}
/**
 * @internal
 */
function createIsParse() {
    halt("createIsParse");
}
/**
 * @internal
 */
function createAssertParse() {
    halt("createAssertParse");
}
/**
 * @internal
 */
function createValidateParse() {
    halt("createValidateParse");
}
/**
 * @internal
 */
function createStringify() {
    halt("createStringify");
}
/**
 * @internal
 */
function createAssertStringify() {
    halt("createAssertStringify");
}
/**
 * @internal
 */
function createIsStringify() {
    halt("createIsStringify");
}
/**
 * @internal
 */
function createValidateStringify() {
    halt("createValidateStringify");
}
/**
 * @internal
 */
function halt(name) {
    throw new Error(`Error on typia.json.${name}(): no transform has been configured. Read and follow https://typia.io/docs/setup please.`);
}
