// typia@7.2.0/lib/module.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tags = exports.reflect = exports.protobuf = exports.notations = exports.misc = exports.json = exports.llm = exports.http = exports.functional = void 0;
exports.assert = assert;
exports.assertGuard = assertGuard;
exports.is = is;
exports.validate = validate;
exports.assertEquals = assertEquals;
exports.assertGuardEquals = assertGuardEquals;
exports.equals = equals;
exports.validateEquals = validateEquals;
exports.random = random;
exports.createAssert = createAssert;
exports.createAssertGuard = createAssertGuard;
exports.createIs = createIs;
exports.createValidate = createValidate;
exports.createAssertEquals = createAssertEquals;
exports.createAssertGuardEquals = createAssertGuardEquals;
exports.createEquals = createEquals;
exports.createValidateEquals = createValidateEquals;
exports.createRandom = createRandom;
exports.functional = __importStar(require("./functional"));
exports.http = __importStar(require("./http"));
exports.llm = __importStar(require("./llm"));
exports.json = __importStar(require("./json"));
exports.misc = __importStar(require("./misc"));
exports.notations = __importStar(require("./notations"));
exports.protobuf = __importStar(require("./protobuf"));
exports.reflect = __importStar(require("./reflect"));
exports.tags = __importStar(require("./tags"));
__exportStar(require("./schemas/metadata/IJsDocTagInfo"), exports);
__exportStar(require("./schemas/json/IJsonApplication"), exports);
__exportStar(require("./schemas/json/IJsonSchemaCollection"), exports);
__exportStar(require("./schemas/llm/ILlmApplicationOfValidate"), exports);
__exportStar(require("./schemas/llm/ILlmFunctionOfValidate"), exports);
__exportStar(require("./AssertionGuard"), exports);
__exportStar(require("./IRandomGenerator"), exports);
__exportStar(require("./IValidation"), exports);
__exportStar(require("./TypeGuardError"), exports);
__exportStar(require("./Primitive"), exports);
__exportStar(require("./Resolved"), exports);
__exportStar(require("./CamelCase"), exports);
__exportStar(require("./PascalCase"), exports);
__exportStar(require("./SnakeCase"), exports);
__exportStar(require("./IReadableURLSearchParams"), exports);
/**
 * @internal
 */
function assert() {
    halt("assert");
}
/**
 * @internal
 */
function assertGuard() {
    halt("assertGuard");
}
/**
 * @internal
 */
function is() {
    halt("is");
}
/**
 * @internal
 */
function validate() {
    halt("validate");
}
/**
 * @internal
 */
function assertEquals() {
    halt("assertEquals");
}
/**
 * @internal
 */
function assertGuardEquals() {
    halt("assertGuardEquals");
}
/**
 * @internal
 */
function equals() {
    halt("equals");
}
/**
 * @internal
 */
function validateEquals() {
    halt("validateEquals");
}
/**
 * @internal
 */
function random() {
    halt("random");
}
/**
 * @internal
 */
function createAssert() {
    halt("createAssert");
}
/**
 * @internal
 */
function createAssertGuard() {
    halt("createAssertGuard");
}
/**
 * @internal
 */
function createIs() {
    halt("createIs");
}
/**
 * @internal
 */
function createValidate() {
    halt("createValidate");
}
/**
 * @internal
 */
function createAssertEquals() {
    halt("createAssertEquals");
}
/**
 * @internal
 */
function createAssertGuardEquals() {
    halt("createAssertGuardEquals");
}
/**
 * @internal
 */
function createEquals() {
    halt("createEquals");
}
/**
 * @internal
 */
function createValidateEquals() {
    halt("createValidateEquals");
}
/**
 * @internal
 */
function createRandom() {
    halt("createRandom");
}
/**
 * @internal
 */
function halt(name) {
    throw new Error(`Error on typia.${name}(): no transform has been configured. Read and follow https://typia.io/docs/setup please.`);
}
