// @samchon/openapi@2.0.4/lib/composers/LlmSchemaComposer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LlmSchemaComposer = void 0;
var ChatGptTypeChecker_1 = require("../utils/ChatGptTypeChecker");
var ClaudeTypeChecker_1 = require("../utils/ClaudeTypeChecker");
var GeminiTypeChecker_1 = require("../utils/GeminiTypeChecker");
var LlamaTypeChecker_1 = require("../utils/LlamaTypeChecker");
var LlmTypeCheckerV3_2 = require("../utils/LlmTypeCheckerV3");
var LlmTypeCheckerV3_1_1 = require("../utils/LlmTypeCheckerV3_1");
var ChatGptSchemaComposer_1 = require("./llm/ChatGptSchemaComposer");
var ClaudeSchemaComposer_1 = require("./llm/ClaudeSchemaComposer");
var GeminiSchemaComposer_1 = require("./llm/GeminiSchemaComposer");
var LlamaSchemaComposer_1 = require("./llm/LlamaSchemaComposer");
var LlmSchemaV3Composer_1 = require("./llm/LlmSchemaV3Composer");
var LlmSchemaV3_1Composer_1 = require("./llm/LlmSchemaV3_1Composer");
var LlmSchemaComposer;
(function (LlmSchemaComposer) {
    LlmSchemaComposer.parameters = function (model) {
        return PARAMETERS_CASTERS[model];
    };
    LlmSchemaComposer.schema = function (model) {
        return SCHEMA_CASTERS[model];
    };
    LlmSchemaComposer.defaultConfig = function (model) {
        return DEFAULT_CONFIGS[model];
    };
    LlmSchemaComposer.typeChecker = function (model) {
        return TYPE_CHECKERS[model];
    };
    LlmSchemaComposer.separateParameters = function (model) { return SEPARATE_PARAMETERS[model]; };
})(LlmSchemaComposer || (exports.LlmSchemaComposer = LlmSchemaComposer = {}));
var PARAMETERS_CASTERS = {
    chatgpt: ChatGptSchemaComposer_1.ChatGptSchemaComposer.parameters,
    claude: ClaudeSchemaComposer_1.ClaudeSchemaComposer.parameters,
    gemini: GeminiSchemaComposer_1.GeminiSchemaComposer.parameters,
    llama: LlamaSchemaComposer_1.LlamaSchemaComposer.parameters,
    "3.0": LlmSchemaV3Composer_1.LlmSchemaV3Composer.parameters,
    "3.1": LlmSchemaV3_1Composer_1.LlmSchemaV3_1Composer.parameters,
};
var SCHEMA_CASTERS = {
    chatgpt: ChatGptSchemaComposer_1.ChatGptSchemaComposer.schema,
    claude: ClaudeSchemaComposer_1.ClaudeSchemaComposer.schema,
    gemini: GeminiSchemaComposer_1.GeminiSchemaComposer.schema,
    llama: LlamaSchemaComposer_1.LlamaSchemaComposer.schema,
    "3.0": LlmSchemaV3Composer_1.LlmSchemaV3Composer.schema,
    "3.1": LlmSchemaV3_1Composer_1.LlmSchemaV3_1Composer.schema,
};
var SEPARATE_PARAMETERS = {
    chatgpt: ChatGptSchemaComposer_1.ChatGptSchemaComposer.separateParameters,
    claude: ClaudeSchemaComposer_1.ClaudeSchemaComposer.separateParameters,
    gemini: GeminiSchemaComposer_1.GeminiSchemaComposer.separateParameters,
    llama: LlamaSchemaComposer_1.LlamaSchemaComposer.separateParameters,
    "3.0": LlmSchemaV3Composer_1.LlmSchemaV3Composer.separateParameters,
    "3.1": LlmSchemaV3_1Composer_1.LlmSchemaV3_1Composer.separateParameters,
};
var DEFAULT_CONFIGS = {
    chatgpt: {
        reference: false,
    },
    claude: {
        reference: false,
    },
    gemini: {
        recursive: 3,
    },
    llama: {
        reference: false,
    },
    "3.0": {
        constraint: true,
        recursive: 3,
    },
    "3.1": {
        constraint: true,
        reference: false,
    },
};
var TYPE_CHECKERS = {
    chatgpt: ChatGptTypeChecker_1.ChatGptTypeChecker,
    claude: ClaudeTypeChecker_1.ClaudeTypeChecker,
    gemini: GeminiTypeChecker_1.GeminiTypeChecker,
    llama: LlamaTypeChecker_1.LlamaTypeChecker,
    "3.0": LlmTypeCheckerV3_2.LlmTypeCheckerV3,
    "3.1": LlmTypeCheckerV3_1_1.LlmTypeCheckerV3_1,
};
