// @samchon/openapi@2.0.4/lib/http/HttpLlmFunctionFetcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpLlmFunctionFetcher = void 0;
var HttpMigrateRouteFetcher_1 = require("./HttpMigrateRouteFetcher");
var HttpLlmFunctionFetcher;
(function (HttpLlmFunctionFetcher) {
    HttpLlmFunctionFetcher.execute = function (props) {
        return HttpMigrateRouteFetcher_1.HttpMigrateRouteFetcher.execute(getFetchArguments("execute", props));
    };
    HttpLlmFunctionFetcher.propagate = function (props) {
        return HttpMigrateRouteFetcher_1.HttpMigrateRouteFetcher.propagate(getFetchArguments("propagate", props));
    };
    var getFetchArguments = function (from, props) {
        var route = props.function.route();
        var input = props.input;
        var valid = typeof input === "object" && input !== null;
        if (valid === false)
            throw new Error("Error on HttpLlmFunctionFetcher.".concat(from, "(): keyworded arguments must be an object"));
        return {
            connection: props.connection,
            route: route,
            parameters: Object.fromEntries(route.parameters.map(function (p) { return [p.key, input[p.key]]; })),
            query: input.query,
            body: input.body,
        };
    };
})(HttpLlmFunctionFetcher || (exports.HttpLlmFunctionFetcher = HttpLlmFunctionFetcher = {}));
