// @samchon/openapi@2.0.4/lib/converters/OpenApiV3Downgrader.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenApiV3Downgrader = void 0;
var OpenApiTypeChecker_1 = require("../utils/OpenApiTypeChecker");
var OpenApiV3Downgrader;
(function (OpenApiV3Downgrader) {
    OpenApiV3Downgrader.downgrade = function (input) {
        var collection = OpenApiV3Downgrader.downgradeComponents(input.components);
        return {
            openapi: "3.0.0",
            servers: input.servers,
            info: input.info,
            components: collection.downgraded,
            paths: input.paths
                ? Object.fromEntries(Object.entries(input.paths)
                    .filter(function (_a) {
                    var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                    return v !== undefined;
                })
                    .map(function (_a) {
                    var _b = __read(_a, 2), key = _b[0], value = _b[1];
                    return [key, downgradePathItem(collection)(value)];
                }))
                : undefined,
            security: input.security,
            tags: input.tags,
        };
    };
    /* -----------------------------------------------------------
      OPERATORS
    ----------------------------------------------------------- */
    var downgradePathItem = function (collection) {
        return function (pathItem) { return (__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign({}, pathItem), (pathItem.get
            ? { get: downgradeOperation(collection)(pathItem.get) }
            : undefined)), (pathItem.put
            ? { put: downgradeOperation(collection)(pathItem.put) }
            : undefined)), (pathItem.post
            ? { post: downgradeOperation(collection)(pathItem.post) }
            : undefined)), (pathItem.delete
            ? { delete: downgradeOperation(collection)(pathItem.delete) }
            : undefined)), (pathItem.options
            ? { options: downgradeOperation(collection)(pathItem.options) }
            : undefined)), (pathItem.head
            ? { head: downgradeOperation(collection)(pathItem.head) }
            : undefined)), (pathItem.patch
            ? { patch: downgradeOperation(collection)(pathItem.patch) }
            : undefined)), (pathItem.trace
            ? { trace: downgradeOperation(collection)(pathItem.trace) }
            : undefined))); };
    };
    var downgradeOperation = function (collection) {
        return function (input) { return (__assign(__assign({}, input), { parameters: input.parameters
                ? input.parameters.map(downgradeParameter(collection))
                : undefined, requestBody: input.requestBody
                ? downgradeRequestBody(collection)(input.requestBody)
                : undefined, responses: input.responses
                ? Object.fromEntries(Object.entries(input.responses)
                    .filter(function (_a) {
                    var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                    return v !== undefined;
                })
                    .map(function (_a) {
                    var _b = __read(_a, 2), key = _b[0], value = _b[1];
                    return [
                        key,
                        downgradeResponse(collection)(value),
                    ];
                }))
                : undefined })); };
    };
    var downgradeParameter = function (collection) {
        return function (input) { return (__assign(__assign({}, input), { schema: OpenApiV3Downgrader.downgradeSchema(collection)(input.schema) })); };
    };
    var downgradeRequestBody = function (collection) {
        return function (input) { return (__assign(__assign({}, input), { content: input.content
                ? downgradeContent(collection)(input.content)
                : undefined })); };
    };
    var downgradeResponse = function (collection) {
        return function (input) { return (__assign(__assign({}, input), { content: input.content
                ? downgradeContent(collection)(input.content)
                : undefined, headers: input.headers
                ? Object.fromEntries(Object.entries(input.headers)
                    .filter(function (_a) {
                    var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                    return v !== undefined;
                })
                    .map(function (_a) {
                    var _b = __read(_a, 2), key = _b[0], value = _b[1];
                    return [
                        key,
                        __assign(__assign({}, value), { schema: OpenApiV3Downgrader.downgradeSchema(collection)(value.schema) }),
                    ];
                }))
                : undefined })); };
    };
    var downgradeContent = function (collection) {
        return function (record) {
            return Object.fromEntries(Object.entries(record)
                .filter(function (_a) {
                var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                return v !== undefined;
            })
                .map(function (_a) {
                var _b = __read(_a, 2), key = _b[0], value = _b[1];
                return [
                    key,
                    __assign(__assign({}, value), { schema: (value === null || value === void 0 ? void 0 : value.schema)
                            ? OpenApiV3Downgrader.downgradeSchema(collection)(value.schema)
                            : undefined }),
                ];
            }));
        };
    };
    /* -----------------------------------------------------------
      DEFINITIONS
    ----------------------------------------------------------- */
    OpenApiV3Downgrader.downgradeComponents = function (input) {
        var e_1, _a;
        var collection = {
            original: input,
            downgraded: {
                securitySchemes: input.securitySchemes,
            },
        };
        if (input.schemas) {
            collection.downgraded.schemas = {};
            try {
                for (var _b = __values(Object.entries(input.schemas)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var _d = __read(_c.value, 2), key = _d[0], value = _d[1];
                    if (value !== undefined)
                        collection.downgraded.schemas[key] =
                            OpenApiV3Downgrader.downgradeSchema(collection)(value);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return collection;
    };
    OpenApiV3Downgrader.downgradeSchema = function (collection) {
        return function (input) {
            var e_2, _a;
            var nullable = isNullable(new Set())(collection.original)(input);
            var union = [];
            var attribute = __assign({ title: input.title, description: input.description, example: input.example, examples: input.examples }, Object.fromEntries(Object.entries(input).filter(function (_a) {
                var _b = __read(_a, 2), key = _b[0], value = _b[1];
                return key.startsWith("x-") && value !== undefined;
            })));
            var visit = function (schema) {
                if (OpenApiTypeChecker_1.OpenApiTypeChecker.isBoolean(schema))
                    union.push({ type: "boolean" });
                else if (OpenApiTypeChecker_1.OpenApiTypeChecker.isBoolean(schema) ||
                    OpenApiTypeChecker_1.OpenApiTypeChecker.isInteger(schema) ||
                    OpenApiTypeChecker_1.OpenApiTypeChecker.isNumber(schema) ||
                    OpenApiTypeChecker_1.OpenApiTypeChecker.isString(schema) ||
                    OpenApiTypeChecker_1.OpenApiTypeChecker.isReference(schema))
                    union.push(__assign({}, schema));
                else if (OpenApiTypeChecker_1.OpenApiTypeChecker.isArray(schema))
                    union.push(__assign(__assign({}, schema), { items: OpenApiV3Downgrader.downgradeSchema(collection)(schema.items) }));
                else if (OpenApiTypeChecker_1.OpenApiTypeChecker.isTuple(schema))
                    union.push(__assign(__assign(__assign({}, schema), { items: (function () {
                            if (schema.additionalItems === true)
                                return {};
                            var elements = __spreadArray(__spreadArray([], __read(schema.prefixItems), false), __read((typeof schema.additionalItems === "object"
                                ? [OpenApiV3Downgrader.downgradeSchema(collection)(schema.additionalItems)]
                                : [])), false);
                            if (elements.length === 0)
                                return {};
                            return {
                                oneOf: elements.map(OpenApiV3Downgrader.downgradeSchema(collection)),
                            };
                        })(), minItems: schema.prefixItems.length, maxItems: !!schema.additionalItems === true
                            ? undefined
                            : schema.prefixItems.length }), {
                        prefixItems: undefined,
                        additionalItems: undefined,
                    }));
                else if (OpenApiTypeChecker_1.OpenApiTypeChecker.isObject(schema))
                    union.push(__assign(__assign({}, schema), { properties: schema.properties
                            ? Object.fromEntries(Object.entries(schema.properties)
                                .filter(function (_a) {
                                var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                                return v !== undefined;
                            })
                                .map(function (_a) {
                                var _b = __read(_a, 2), key = _b[0], value = _b[1];
                                return [
                                    key,
                                    OpenApiV3Downgrader.downgradeSchema(collection)(value),
                                ];
                            }))
                            : undefined, additionalProperties: typeof schema.additionalProperties === "object"
                            ? OpenApiV3Downgrader.downgradeSchema(collection)(schema.additionalProperties)
                            : schema.additionalProperties, required: schema.required }));
                else if (OpenApiTypeChecker_1.OpenApiTypeChecker.isOneOf(schema))
                    schema.oneOf.forEach(visit);
            };
            var visitConstant = function (schema) {
                var e_3, _a;
                var insert = function (value) {
                    var _a;
                    var matched = union.find(function (u) {
                        return u.type ===
                            typeof value;
                    });
                    if (matched !== undefined) {
                        (_a = matched.enum) !== null && _a !== void 0 ? _a : (matched.enum = []);
                        matched.enum.push(value);
                    }
                    else
                        union.push({ type: typeof value, enum: [value] });
                };
                if (OpenApiTypeChecker_1.OpenApiTypeChecker.isConstant(schema))
                    insert(schema.const);
                else if (OpenApiTypeChecker_1.OpenApiTypeChecker.isOneOf(schema))
                    try {
                        for (var _b = __values(schema.oneOf), _c = _b.next(); !_c.done; _c = _b.next()) {
                            var u = _c.value;
                            if (OpenApiTypeChecker_1.OpenApiTypeChecker.isConstant(u))
                                insert(u.const);
                        }
                    }
                    catch (e_3_1) { e_3 = { error: e_3_1 }; }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                        }
                        finally { if (e_3) throw e_3.error; }
                    }
            };
            visit(input);
            visitConstant(input);
            if (nullable === true)
                try {
                    for (var union_1 = __values(union), union_1_1 = union_1.next(); !union_1_1.done; union_1_1 = union_1.next()) {
                        var u = union_1_1.value;
                        if (OpenApiTypeChecker_1.OpenApiTypeChecker.isReference(u))
                            downgradeNullableReference(new Set())(collection)(u);
                        else
                            u.nullable = true;
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (union_1_1 && !union_1_1.done && (_a = union_1.return)) _a.call(union_1);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
            if (nullable === true && union.length === 0)
                return __assign({ type: "null" }, attribute);
            return __assign(__assign({}, (union.length === 0
                ? { type: undefined }
                : union.length === 1
                    ? __assign({}, union[0]) : { oneOf: union })), attribute);
        };
    };
    var downgradeNullableReference = function (visited) {
        return function (collection) {
            return function (schema) {
                var _a, _b, _c;
                var _d;
                var key = schema.$ref.split("/").pop();
                if (key.endsWith(".Nullable"))
                    return;
                var found = (_a = collection.original.schemas) === null || _a === void 0 ? void 0 : _a[key];
                if (found === undefined)
                    return;
                else if (isNullable(visited)(collection.original)(found) === true)
                    return;
                else if (((_b = collection.downgraded.schemas) === null || _b === void 0 ? void 0 : _b["".concat(key, ".Nullable")]) === undefined) {
                    (_c = (_d = collection.downgraded).schemas) !== null && _c !== void 0 ? _c : (_d.schemas = {});
                    collection.downgraded.schemas["".concat(key, ".Nullable")] = {};
                    collection.downgraded.schemas["".concat(key, ".Nullable")] = OpenApiV3Downgrader.downgradeSchema(collection)(OpenApiTypeChecker_1.OpenApiTypeChecker.isOneOf(found)
                        ? __assign(__assign({}, found), { oneOf: __spreadArray(__spreadArray([], __read(found.oneOf), false), [{ type: "null" }], false) }) : __assign({ oneOf: [found, { type: "null" }], title: found.title, description: found.description, example: found.example, examples: found.examples }, Object.fromEntries(Object.entries(found).filter(function (_a) {
                        var _b = __read(_a, 2), key = _b[0], value = _b[1];
                        return key.startsWith("x-") && value !== undefined;
                    }))));
                }
                schema.$ref += ".Nullable";
            };
        };
    };
    var isNullable = function (visited) {
        return function (components) {
            return function (schema) {
                var _a;
                if (OpenApiTypeChecker_1.OpenApiTypeChecker.isNull(schema))
                    return true;
                else if (OpenApiTypeChecker_1.OpenApiTypeChecker.isReference(schema)) {
                    if (visited.has(schema.$ref))
                        return false;
                    visited.add(schema.$ref);
                    var key = schema.$ref.split("/").pop();
                    var next = (_a = components.schemas) === null || _a === void 0 ? void 0 : _a[key];
                    return next ? isNullable(visited)(components)(next) : false;
                }
                return (OpenApiTypeChecker_1.OpenApiTypeChecker.isOneOf(schema) &&
                    schema.oneOf.some(isNullable(visited)(components)));
            };
        };
    };
})(OpenApiV3Downgrader || (exports.OpenApiV3Downgrader = OpenApiV3Downgrader = {}));
