// @samchon/openapi@2.0.4/lib/converters/OpenApiV3Upgrader.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenApiV3Upgrader = void 0;
var OpenApiTypeChecker_1 = require("../utils/OpenApiTypeChecker");
var OpenApiV3Upgrader;
(function (OpenApiV3Upgrader) {
    OpenApiV3Upgrader.convert = function (input) {
        var _a;
        return (__assign(__assign({}, input), { components: OpenApiV3Upgrader.convertComponents((_a = input.components) !== null && _a !== void 0 ? _a : {}), paths: input.paths
                ? Object.fromEntries(Object.entries(input.paths)
                    .filter(function (_a) {
                    var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                    return v !== undefined;
                })
                    .map(function (_a) {
                    var _b = __read(_a, 2), key = _b[0], value = _b[1];
                    return [key, convertPathItem(input)(value)];
                }))
                : undefined, openapi: "3.1.0", "x-samchon-emended": true }));
    };
    /* -----------------------------------------------------------
      OPERATORS
    ----------------------------------------------------------- */
    var convertPathItem = function (doc) {
        return function (pathItem) { return (__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign({}, pathItem), (pathItem.get
            ? { get: convertOperation(doc)(pathItem)(pathItem.get) }
            : undefined)), (pathItem.put
            ? { put: convertOperation(doc)(pathItem)(pathItem.put) }
            : undefined)), (pathItem.post
            ? { post: convertOperation(doc)(pathItem)(pathItem.post) }
            : undefined)), (pathItem.delete
            ? { delete: convertOperation(doc)(pathItem)(pathItem.delete) }
            : undefined)), (pathItem.options
            ? { options: convertOperation(doc)(pathItem)(pathItem.options) }
            : undefined)), (pathItem.head
            ? { head: convertOperation(doc)(pathItem)(pathItem.head) }
            : undefined)), (pathItem.patch
            ? { patch: convertOperation(doc)(pathItem)(pathItem.patch) }
            : undefined)), (pathItem.trace
            ? { trace: convertOperation(doc)(pathItem)(pathItem.trace) }
            : undefined))); };
    };
    var convertOperation = function (doc) {
        return function (pathItem) {
            return function (input) {
                var _a, _b;
                return (__assign(__assign({}, input), { parameters: pathItem.parameters !== undefined || input.parameters !== undefined
                        ? __spreadArray(__spreadArray([], __read(((_a = pathItem.parameters) !== null && _a !== void 0 ? _a : [])), false), __read(((_b = input.parameters) !== null && _b !== void 0 ? _b : [])), false).map(function (p) {
                            var _a, _b, _c, _d, _e, _f, _g, _h;
                            if (!TypeChecker.isReference(p))
                                return convertParameter((_a = doc.components) !== null && _a !== void 0 ? _a : {})(p);
                            var found = p.$ref.startsWith("#/components/headers/")
                                ? (_c = (_b = doc.components) === null || _b === void 0 ? void 0 : _b.headers) === null || _c === void 0 ? void 0 : _c[(_d = p.$ref.split("/").pop()) !== null && _d !== void 0 ? _d : ""]
                                : (_f = (_e = doc.components) === null || _e === void 0 ? void 0 : _e.parameters) === null || _f === void 0 ? void 0 : _f[(_g = p.$ref.split("/").pop()) !== null && _g !== void 0 ? _g : ""];
                            return found !== undefined
                                ? convertParameter((_h = doc.components) !== null && _h !== void 0 ? _h : {})(__assign(__assign({}, found), { in: "header" }))
                                : undefined;
                        })
                            .filter(function (_, v) { return v !== undefined; })
                        : undefined, requestBody: input.requestBody
                        ? convertRequestBody(doc)(input.requestBody)
                        : undefined, responses: input.responses
                        ? Object.fromEntries(Object.entries(input.responses)
                            .filter(function (_a) {
                            var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                            return v !== undefined;
                        })
                            .map(function (_a) {
                            var _b = __read(_a, 2), key = _b[0], value = _b[1];
                            return [key, convertResponse(doc)(value)];
                        })
                            .filter(function (_a) {
                            var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                            return v !== undefined;
                        }))
                        : undefined }));
            };
        };
    };
    var convertParameter = function (components) {
        return function (input) { return (__assign(__assign({}, input), { schema: OpenApiV3Upgrader.convertSchema(components)(input.schema), examples: input.examples
                ? Object.fromEntries(Object.entries(input.examples)
                    .map(function (_a) {
                    var _b, _c;
                    var _d = __read(_a, 2), key = _d[0], value = _d[1];
                    return [
                        key,
                        TypeChecker.isReference(value)
                            ? (_b = components.examples) === null || _b === void 0 ? void 0 : _b[(_c = value.$ref.split("/").pop()) !== null && _c !== void 0 ? _c : ""]
                            : value,
                    ];
                })
                    .filter(function (_a) {
                    var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                    return v !== undefined;
                }))
                : undefined })); };
    };
    var convertRequestBody = function (doc) {
        return function (input) {
            var _a, _b, _c, _d;
            if (TypeChecker.isReference(input)) {
                var found = (_b = (_a = doc.components) === null || _a === void 0 ? void 0 : _a.requestBodies) === null || _b === void 0 ? void 0 : _b[(_c = input.$ref.split("/").pop()) !== null && _c !== void 0 ? _c : ""];
                if (found === undefined)
                    return undefined;
                input = found;
            }
            return __assign(__assign({}, input), { content: input.content
                    ? convertContent((_d = doc.components) !== null && _d !== void 0 ? _d : {})(input.content)
                    : undefined });
        };
    };
    var convertResponse = function (doc) {
        return function (input) {
            var _a, _b, _c, _d;
            if (TypeChecker.isReference(input)) {
                var found = (_b = (_a = doc.components) === null || _a === void 0 ? void 0 : _a.responses) === null || _b === void 0 ? void 0 : _b[(_c = input.$ref.split("/").pop()) !== null && _c !== void 0 ? _c : ""];
                if (found === undefined)
                    return undefined;
                input = found;
            }
            return __assign(__assign({}, input), { content: input.content
                    ? convertContent((_d = doc.components) !== null && _d !== void 0 ? _d : {})(input.content)
                    : undefined, headers: input.headers
                    ? Object.fromEntries(Object.entries(input.headers)
                        .filter(function (_a) {
                        var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                        return v !== undefined;
                    })
                        .map(function (_a) {
                        var _b = __read(_a, 2), key = _b[0], value = _b[1];
                        return [
                            key,
                            (function () {
                                var _a, _b, _c, _d, _e;
                                if (TypeChecker.isReference(value) === false)
                                    return convertParameter((_a = doc.components) !== null && _a !== void 0 ? _a : {})(__assign(__assign({}, value), { in: "header" }));
                                var found = value.$ref.startsWith("#/components/headers/")
                                    ? (_c = (_b = doc.components) === null || _b === void 0 ? void 0 : _b.headers) === null || _c === void 0 ? void 0 : _c[(_d = value.$ref.split("/").pop()) !== null && _d !== void 0 ? _d : ""]
                                    : undefined;
                                return found !== undefined
                                    ? convertParameter((_e = doc.components) !== null && _e !== void 0 ? _e : {})(__assign(__assign({}, found), { in: "header" }))
                                    : undefined;
                            })(),
                        ];
                    })
                        .filter(function (_a) {
                        var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                        return v !== undefined;
                    }))
                    : undefined });
        };
    };
    var convertContent = function (components) {
        return function (record) {
            return Object.fromEntries(Object.entries(record)
                .filter(function (_a) {
                var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                return v !== undefined;
            })
                .map(function (_a) {
                var _b = __read(_a, 2), key = _b[0], value = _b[1];
                return [
                    key,
                    __assign(__assign({}, value), { schema: value.schema
                            ? OpenApiV3Upgrader.convertSchema(components)(value.schema)
                            : undefined, examples: value.examples
                            ? Object.fromEntries(Object.entries(value.examples)
                                .map(function (_a) {
                                var _b, _c;
                                var _d = __read(_a, 2), key = _d[0], value = _d[1];
                                return [
                                    key,
                                    TypeChecker.isReference(value)
                                        ? (_b = components.examples) === null || _b === void 0 ? void 0 : _b[(_c = value.$ref.split("/").pop()) !== null && _c !== void 0 ? _c : ""]
                                        : value,
                                ];
                            })
                                .filter(function (_a) {
                                var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                                return v !== undefined;
                            }))
                            : undefined }),
                ];
            }));
        };
    };
    /* -----------------------------------------------------------
      DEFINITIONS
    ----------------------------------------------------------- */
    OpenApiV3Upgrader.convertComponents = function (input) {
        var _a;
        return ({
            schemas: Object.fromEntries(Object.entries((_a = input.schemas) !== null && _a !== void 0 ? _a : {})
                .filter(function (_a) {
                var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                return v !== undefined;
            })
                .map(function (_a) {
                var _b = __read(_a, 2), key = _b[0], value = _b[1];
                return [key, OpenApiV3Upgrader.convertSchema(input)(value)];
            })),
            securitySchemes: input.securitySchemes,
        });
    };
    OpenApiV3Upgrader.convertSchema = function (components) {
        return function (input) {
            var e_1, _a;
            var nullable = {
                value: false,
                default: undefined,
            };
            var union = [];
            var attribute = __assign(__assign({ title: input.title, description: input.description }, Object.fromEntries(Object.entries(input).filter(function (_a) {
                var _b = __read(_a, 2), key = _b[0], value = _b[1];
                return key.startsWith("x-") && value !== undefined;
            }))), { example: input.example, examples: input.examples });
            var visit = function (schema) {
                var _a, _b, _c, _d, _e;
                // NULLABLE PROPERTY
                if (schema.nullable ===
                    true) {
                    nullable.value || (nullable.value = true);
                    if (schema.default === null)
                        nullable.default = null;
                }
                if (Array.isArray(schema.enum) &&
                    ((_a = schema.enum) === null || _a === void 0 ? void 0 : _a.length) &&
                    ((_b = schema.enum) === null || _b === void 0 ? void 0 : _b.some(function (e) { return e === null; })))
                    nullable.value || (nullable.value = true);
                // UNION TYPE CASE
                if (TypeChecker.isAnyOf(schema))
                    schema.anyOf.forEach(visit);
                else if (TypeChecker.isOneOf(schema))
                    schema.oneOf.forEach(visit);
                else if (TypeChecker.isAllOf(schema))
                    if (schema.allOf.length === 1)
                        visit(schema.allOf[0]);
                    else
                        union.push(convertAllOfSchema(components)(schema));
                // ATOMIC TYPE CASE (CONSIDER ENUM VALUES)
                else if (TypeChecker.isBoolean(schema) ||
                    TypeChecker.isInteger(schema) ||
                    TypeChecker.isNumber(schema) ||
                    TypeChecker.isString(schema))
                    if (((_c = schema.enum) === null || _c === void 0 ? void 0 : _c.length) &&
                        schema.enum.filter(function (e) { return e !== null; }).length)
                        union.push.apply(union, __spreadArray([], __read(schema.enum
                            .filter(function (v) { return v !== null; })
                            .map(function (value) { return ({ const: value }); })), false));
                    else
                        union.push(__assign(__assign(__assign({}, schema), { default: ((_d = schema.default) !== null && _d !== void 0 ? _d : undefined) }), { enum: undefined }));
                // INSTANCE TYPE CASE
                else if (TypeChecker.isArray(schema))
                    union.push(__assign(__assign({}, schema), { items: OpenApiV3Upgrader.convertSchema(components)(schema.items) }));
                else if (TypeChecker.isObject(schema))
                    union.push(__assign(__assign({}, schema), {
                        properties: schema.properties
                            ? Object.fromEntries(Object.entries(schema.properties)
                                .filter(function (_a) {
                                var _b = __read(_a, 2), _ = _b[0], v = _b[1];
                                return v !== undefined;
                            })
                                .map(function (_a) {
                                var _b = __read(_a, 2), key = _b[0], value = _b[1];
                                return [
                                    key,
                                    OpenApiV3Upgrader.convertSchema(components)(value),
                                ];
                            }))
                            : {},
                        additionalProperties: schema.additionalProperties
                            ? typeof schema.additionalProperties === "object" &&
                                schema.additionalProperties !== null
                                ? OpenApiV3Upgrader.convertSchema(components)(schema.additionalProperties)
                                : schema.additionalProperties
                            : undefined,
                        required: (_e = schema.required) !== null && _e !== void 0 ? _e : [],
                    }));
                else if (TypeChecker.isReference(schema))
                    union.push(schema);
                else
                    union.push(schema);
            };
            visit(input);
            if (nullable.value === true &&
                !union.some(function (e) { return e.type === "null"; }))
                union.push({
                    type: "null",
                    default: nullable.default,
                });
            if (union.length === 2 &&
                union.filter(function (x) { return OpenApiTypeChecker_1.OpenApiTypeChecker.isNull(x); }).length === 1) {
                var type = union.filter(function (x) { return OpenApiTypeChecker_1.OpenApiTypeChecker.isNull(x) === false; })[0];
                try {
                    for (var _b = __values([
                        "title",
                        "description",
                        "deprecated",
                        "example",
                        "examples",
                    ]), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var key = _c.value;
                        if (type[key] !== undefined)
                            delete type[key];
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            return __assign(__assign(__assign({}, (union.length === 0
                ? { type: undefined }
                : union.length === 1
                    ? __assign({}, union[0]) : { oneOf: union.map(function (u) { return (__assign(__assign({}, u), { nullable: undefined })); }) })), attribute), { nullable: undefined });
        };
    };
    var convertAllOfSchema = function (components) {
        return function (input) {
            var objects = input.allOf.map(function (schema) { return retrieveObject(components)(schema); });
            if (objects.some(function (obj) { return obj === null; }))
                return __assign({ type: undefined }, {
                    allOf: undefined,
                });
            return __assign(__assign(__assign({}, input), { type: "object", properties: Object.fromEntries(objects
                    .map(function (o) { var _a; return Object.entries((_a = o === null || o === void 0 ? void 0 : o.properties) !== null && _a !== void 0 ? _a : {}); })
                    .flat()
                    .map(function (_a) {
                    var _b = __read(_a, 2), key = _b[0], value = _b[1];
                    return [key, OpenApiV3Upgrader.convertSchema(components)(value)];
                })) }), {
                allOf: undefined,
                required: __spreadArray([], __read(new Set(objects.map(function (o) { var _a; return (_a = o === null || o === void 0 ? void 0 : o.required) !== null && _a !== void 0 ? _a : []; }).flat())), false),
            });
        };
    };
    var retrieveObject = function (components) {
        return function (input, visited) {
            var _a, _b, _c;
            if (visited === void 0) { visited = new Set(); }
            if (TypeChecker.isObject(input))
                return input.properties !== undefined && !input.additionalProperties
                    ? input
                    : null;
            else if (visited.has(input))
                return null;
            else
                visited.add(input);
            if (TypeChecker.isReference(input))
                return retrieveObject(components)((_c = (_a = components.schemas) === null || _a === void 0 ? void 0 : _a[(_b = input.$ref.split("/").pop()) !== null && _b !== void 0 ? _b : ""]) !== null && _c !== void 0 ? _c : {}, visited);
            return null;
        };
    };
    var TypeChecker;
    (function (TypeChecker) {
        TypeChecker.isBoolean = function (schema) {
            return schema.type === "boolean";
        };
        TypeChecker.isInteger = function (schema) {
            return schema.type === "integer";
        };
        TypeChecker.isNumber = function (schema) {
            return schema.type === "number";
        };
        TypeChecker.isString = function (schema) {
            return schema.type === "string";
        };
        TypeChecker.isArray = function (schema) {
            return schema.type === "array";
        };
        TypeChecker.isObject = function (schema) {
            return schema.type === "object";
        };
        TypeChecker.isReference = function (schema) {
            return schema.$ref !== undefined;
        };
        TypeChecker.isAllOf = function (schema) {
            return schema.allOf !== undefined;
        };
        TypeChecker.isAnyOf = function (schema) {
            return schema.anyOf !== undefined;
        };
        TypeChecker.isOneOf = function (schema) {
            return schema.oneOf !== undefined;
        };
        TypeChecker.isNullOnly = function (schema) {
            return schema.type === "null";
        };
    })(TypeChecker = OpenApiV3Upgrader.TypeChecker || (OpenApiV3Upgrader.TypeChecker = {}));
})(OpenApiV3Upgrader || (exports.OpenApiV3Upgrader = OpenApiV3Upgrader = {}));
