// @samchon/openapi@2.0.4/lib/composers/migrate/MigrateConverter.mjs
import { EndpointUtil } from "../../utils/EndpointUtil.mjs";

import { HttpMigrateApplicationComposer } from "../HttpMigrateApplicationComposer.mjs";

import { MigrateRouteAccessor } from "./MigrateRouteAccessor.mjs";

var MigrateConverter;

(function(MigrateConverter) {
    MigrateConverter.convert = document => {
        const errors = [];
        const entire = Object.entries({
            ...document.paths ?? {},
            ...document.webhooks ?? {}
        }).map((([path, collection]) => [ "head", "get", "post", "put", "patch", "delete" ].filter((method => collection[method] !== undefined)).map((method => {
            const operation = collection[method];
            const migrated = HttpMigrateApplicationComposer.application({
                document,
                method,
                path,
                emendedPath: EndpointUtil.reJoinWithDecimalParameters(path),
                operation
            });
            if (Array.isArray(migrated)) {
                errors.push({
                    method,
                    path,
                    operation: () => operation,
                    messages: migrated
                });
                return null;
            }
            return migrated;
        })))).flat();
        const operations = entire.filter((o => !!o));
        MigrateRouteAccessor.overwrite(operations);
        return {
            document: () => document,
            routes: operations,
            errors
        };
    };
})(MigrateConverter || (MigrateConverter = {}));

export { MigrateConverter };
